Library/Homebrew/formulary.rb

```diff
@@ -191,6 +193,27 @@ def self.from_rack(rack, spec=:stable)
   end
 
   def self.canonical_name(ref)
+    case ref
+    when %r{^([\w+-.]+)$}
+      return ref.downcase if (HOMEBREW_CELLAR/ref.downcase).directory?
+      if (link = (HOMEBREW_CELLAR/ref.downcase).resolved_path).directory?
+        basename = link.basename
+        if FORMULA_RENAMES[ref.downcase] == basename
+          return basename
+        end
+
+        if Tap.detect { |tap| tap.formula_renames[ref.downcase] == basename }
+          return basename
+        end
+      end
+    when HOMEBREW_TAP_FORMULA_REGEX
+      user, repo, name = ref.split("/", 3).map(&:downcase)
+      tap = Tap.new user, repo.sub(/^homebrew-/, "")
+      if (link = (HOMEBREW_CELLAR/name).resolved_path).directory?
+        return basename if tap.formula_renames[name] == link.basename
+      end
+    end
+
```

Comment:
I implemented it as didn't see any reply to [this comment](https://github.com/Homebrew/homebrew/pull/41006#issuecomment-126702686) and thought you're OK with that.

The problem here is that we need to somehow refer to kegs that e.g. installed from path. If I install `ack` from path and core `ack` gets renamed to `newack` I can't uninstall that installed keg, I can't link/unlink it.